### PR TITLE
Detaching from Persistent Economy

### PR DESCRIPTION
**Pull request by SirRichardFrancis**
## About The Pull Request

Does what #2106 says, with an exception of adjusting available item lists and prices, same with mining vendor, it only contains couple of power nodes right now.
Haven't touched the balancing part, since I'm barely familiar with the build and it's IC economy.

### CODE REVIEW

File: code/__defines/research.dm [hunk 1]
```diff
@@ -12,10 +12,11 @@
...
-#define STORE		32	//Purchasing in store
\ No newline at end of file
+#define NONE				0x1
+#define IMPRINTER			0x2		// For circuits. Uses glass/chemicals.
+#define PROTOLATHE			0x4		// New stuff. Uses glass/metal/chemicals
+#define MECHFAB				0x8		// Mechfab
+#define CHASSIS				0x10	// For protolathe, but differently
+#define AUTOLATHE			0x20	// For autolathe
+#define STORE_SCHEMATICS	0x40	// Bound to schematic, unlockable in store with said chematic
+#define STORE_ROUNDSTART	0x80	// Always available in store
```
Comment (DTraitor) [inline code/__defines/research.dm]: use `(1<<X)` instead

File: code/controllers/subsystems/ticker.dm [hunk 2]
```diff
@@ -188,6 +187,8 @@ SUBSYSTEM_DEF(ticker)
 	log_world("Game start took [(world.timeofday - init_start) / 10]s")
 	round_start_time = world.time
 
+	GLOB.using_map.setup_economy()
```
Comment (DTraitor) [inline code/controllers/subsystems/ticker.dm]: Remove this one, I've added `setup_economy()` to `/datum/map/proc/setup_map()` some time ago

File: code/modules/economy/credits.dm [hunk 3]
```diff
@@ -30,7 +22,7 @@
 /obj/item/weapon/spacecash/ewallet/update_icon()
 	icon_state = "grey"
 	switch(worth)
-		if (1 to 500)
+		if (0 to 500)
```
Comment (DTraitor) [inline code/modules/economy/credits.dm]: It should be `icon_state = "grey"` if there are no credits

File: code/modules/economy/store/store_credits.dm [hunk 4]
```diff
@@ -23,49 +31,52 @@
...
+		var/pay_amount = min(occupant.wearing_rig.get_account_balance(), left_to_pay)
 		var/obj/item/weapon/rig/R = occupant.wearing_rig
 		R.charge_to_rig_account(src, "Store Purchase", machine_id, -pay_amount)
-		cost -= pay_amount
+		left_to_pay -= pay_amount
 
+	if(left_to_pay)
+		for(var/i in deposit_box.contents)
+			if(istype(i, /obj/item/weapon/spacecash))
+				var/obj/item/weapon/spacecash/S = i
```
Comment (DTraitor) [inline code/modules/economy/store/store_credits.dm:56]: Use `for(var/obj/item/weapon/spacecash/S in deposit_box.contents)` instead, for loop + istype is faster only for short types (according to my tests)

File: code/modules/economy/store/store_credits.dm [hunk 5]
```diff
@@ -23,49 +31,52 @@
...
+					S.worth -= left_to_pay
+					S.update_icon()
+					deposit_box.update_ui_data()
+					return TRUE
 
-	if (cost <= 0)
-		return TRUE
+	var/datum/money_account/A = occupant?.get_account()
+	if(A)
+		A.money -= left_to_pay
```
Comment (DTraitor) [inline code/modules/economy/store/store_credits.dm:69]: There are no safety checks here meaning our account balance can become lower than 0

File: code/modules/economy/store/store_credits.dm [hunk 6]
```diff
@@ -115,8 +126,10 @@
 /obj/machinery/store/proc/buy_and_transfer()
 	var/list/things = list() + buy_current()
 	var/obj/item/weapon/rig/R
-	for (var/obj/item/I in things)
-		if (isrig(I))
+	if(occupant.back && !occupant.wearing_rig)
+		store_or_drop(occupant.back)
+	for(var/obj/item/I in things)
+		if(isrig(I))
 			R = I
```
Comment (DTraitor) [inline code/modules/economy/store/store_credits.dm]: Just use `var/obj/item/weapon/rig/R = locate(/obj/item/weapon/rig) in things` instead of for loop

**Review by DTraitor — CHANGES_REQUESTED** [review]
Just a few tweaks

File: code/modules/economy/store/store_credits.dm (hunk 5 above)
Comment (SirRichardFrancis) [reply code/modules/economy/store/store_credits.dm:69]: We get here only if `occupant_can_afford()` returned `TRUE`, meaning user have enough money, but no money sources other than this one left, hence check here is redundant, as there is no cases for 'not enough money' at this point.

File: code/modules/economy/store/store_credits.dm (hunk 6 above)
Comment (SirRichardFrancis) [reply code/modules/economy/store/store_credits.dm]: There is `store_or_drop(I)` called for every item in `things`, although I'm not sure if there is a situation when that list contain more than one object.

Comment (DTraitor) [reply code/modules/economy/store/store_credits.dm]: In that case use `locate()` then call `start_transfer(R)` and remove it from `things`, after that we can loop through all items in `things` without checking if it's a RIG.

**Review by DTraitor — APPROVED** [review]
Looks good now

**Pull request merged**